src/lib/optimizer/strategy/predicate_split_up_rule.cpp

```diff
@@ -29,18 +27,81 @@ void PredicateSplitUpRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root
+   *      manipulating the LQP within `visit_lqp()`, while theoretically possible, is prone to bugs.
    */
   for (const auto& [predicate_node, flat_conjunction] : predicate_nodes_to_flat_conjunctions) {
     for (const auto& predicate_expression : flat_conjunction) {
       lqp_insert_node(predicate_node, LQPInputSide::Left, PredicateNode::make(predicate_expression));
     }
     lqp_remove_node(predicate_node);
   }
+
+  return true;
+}
+
+bool PredicateSplitUpRule::splitOr(const std::shared_ptr<AbstractLQPNode>& node) const {
+  /**
+   * Step 1:
+   *    - Collect PredicateNodes that can be split up into multiple ones into `predicate_nodes_to_flat_disjunctions`
+   */
+  auto predicate_nodes_to_flat_disjunctions =
+      std::vector<std::pair<std::shared_ptr<PredicateNode>, std::vector<std::shared_ptr<AbstractExpression>>>>{};
+
+  visit_lqp(node, [&](const auto& sub_node) {
+    if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
+      const auto flat_disjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::Or);
+
+      if (flat_disjunction.size() > 1) {
+        predicate_nodes_to_flat_disjunctions.emplace_back(predicate_node, flat_disjunction);
+      }
+    }
+
+    return LQPVisitation::VisitInputs;
+  });
+
+  if (predicate_nodes_to_flat_disjunctions.empty()) {
+    return false;
+  }
+
+  /**
+   * Step 2:
+   *    - Split up PredicateNodes with disjunctive chains (e.g., `PredicateNode(a OR b OR c)`) as their scan expression
+   *      into n-1 consecutive UnionNodes and n PredicateNodes. We have to do this in a second pass because manipulating
+   *      the LQP within `visit_lqp()`, while theoretically possible, is prone to bugs.
+   */
+  for (const auto& [predicate_node, flat_disjunction] : predicate_nodes_to_flat_disjunctions) {
+    auto previous_union_node = UnionNode::make(UnionMode::Positions);
+    const auto left_input = predicate_node->left_input();
+    lqp_replace_node(predicate_node, previous_union_node);
+    previous_union_node->set_left_input(PredicateNode::make(flat_disjunction[0], left_input));
+    previous_union_node->set_right_input(PredicateNode::make(flat_disjunction[1], left_input));
+
+    for (auto disjunction_idx = size_t{2}; disjunction_idx < flat_disjunction.size(); ++disjunction_idx) {
+      const auto& predicate_expression = flat_disjunction[disjunction_idx];
+      auto next_union_node = UnionNode::make(UnionMode::Positions);
+      lqp_insert_node(previous_union_node, LQPInputSide::Right, next_union_node);
+      next_union_node->set_right_input(PredicateNode::make(predicate_expression, left_input));
+      previous_union_node = next_union_node;
+    }
+  }
+
+  return true;
+}
+
+void PredicateSplitUpRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  Assert(root->type == LQPNodeType::Root, "PredicateSplitUpRule needs root to hold onto");
+
+  while (splitAnd(root) || splitOr(root)) {
```

Comment:
Keep in mind that this is doing redundant re-traversals of the entire LQP. As soon as a single splitable predicate is there, this is doing four(!) traversals over the entire LQP.

Reply:
How much do we care about the optimizer's performance?

Reply:
A change here could come at the cost of clarity.

Reply:
Ask the dude who does TPCC #1803 

It's not just performance, though. Code should just do what is necessary for the sake of the person trying to understand it while stepping/breaking through it in a debugger, or looking at it via a profiler capture.

Reply:
Also, think of a person who's tasked to implement a similar rule and adapts this rule. They might be confused by the apparent redundancy and might deduce by implication that it is necessary for their rule as well, even if good sense tells them otherwise.

Reply:
Alright, I will change it.

Reply:
If you find a good way. I just said "keep in mind..." ;)